### DISCUSSION TITLE
Server -> Instances -> Databases: Memory and Data Path Configuration/Settings

### DISCUSSION THREAD

**Discussion by athanhat** [geldata/gel#4298] [Q&A] [answered]
Hi, I have seriously started looking at EdgeDB to use it in one of my projects and I am planning to run benchmarks and tests on several datasets in order to see how it behaves. This is my environment:

### Environment 

OS: Ubuntu 20.04

``` bash
Installation
curl https://sh.edgedb.com --proto '=https' -sSf1 | sh


edgedb server info --latest
┌─────────────┬────────────────────────────────────────────────────────────────┐
│ Version     │ 2.1+c73716f                                                    │
│ Binary path │ /home/athan/.local/share/edgedb/portable/2.1/bin/edgedb-server │
└─────────────┴────────────────────────────────────────────────────────────────┘

edgedb info
EdgeDB uses the following local paths:
┌────────────┬───────────────────────────────────────┐
│ Cache      │ /home/athan/.cache/edgedb/            │
│ Config     │ /home/athan/.config/edgedb/           │
│ CLI Binary │ /home/athan/.local/bin/edgedb         │
│ Data       │ /home/athan/.local/share/edgedb/data/ │
│ Service    │ /home/athan/.config/systemd/user/     │
└────────────┴───────────────────────────────────────┘
```

### Questions
``` bash
edgedb instance list
┌───────┬────────────┬───────┬─────────────┬─────────┐
│ Kind  │ Name       │ Port  │ Version     │ Status  │
├───────┼────────────┼───────┼─────────────┼─────────┤
│ local │ movies     │ 10702 │ 2.1+c73716f │ running │
│ local │ quickstart │ 10700 │ 2.1+c73716f │ running │
└───────┴────────────┴───────┴─────────────┴─────────┘

edgedb instance status -I quickstart
Running, pid 1228518

edgedb instance status -I movies
Running, pid 1228031
```

1. I created two instances executing `edgedb project init`
  i. the first one running on 10700 (UI) did not ask for login credentials
  ii. the second one running on 10702 (UI) displayed login/password fields but I was not asked to set credentials. 
  iii. what is the default password <hidden> ?
  I couldn't find it so I had to do 
`edgedb instance reset-password --instance movies --user edgedb --password`

``` bash
 edgedb instance credentials 
┌──────────────┬───────────┐
│ Host         │ localhost │
│ Port         │ 10702     │
│ User         │ edgedb    │
│ Password     │ <hidden>  │
│ Database     │ edgedb    │
│ TLS Security │ Default   │
└──────────────┴───────────┘
```
 2. Any of those two instances if it stays inactive for some time, it is automatically shutdown
```
`edgedb instance status -I movies`
Ready in socket activation mode, not running


 ps -Flww -p 1228518
/home/athan/.local/share/edgedb/portable/2.1/bin/python3 
-I /home/athan/.local/share/edgedb/portable/2.1/bin/edgedb-server.py 
--data-dir /home/athan/.local/share/edgedb/data/quickstart 
--runstate-dir /run/user/1000/edgedb-quickstart 
--port 10700 
--compiler-pool-mode=on_demand 
--admin-ui=enabled 
--auto-shutdown-after=600
```
i. How can I change the path of the data-dir for the instance ?
ii. How can I change the auto-shutdown-after ?
iii. What is the analogy of instances with Postgres here ?
Is this instance a Postgres process that is forked from the server ?
```sql
select datname, pid, usename, application_name, state, client_port 
from pg_stat_activity;
```

3. Especially for the data-dir I have the same question but for the `edgedb info` (server)
It appears that this is the edgedb server location where all instances save data `/home/athan/.local/share/edgedb/data/`
Is it possible to relocate data to a different storage device and/or specify different data folder for the server ?
 
4. edgedb -> instances -> databases
i. I am trying to understand where is the Postgres server located in your architecture.
This is what I discovered from the logs:
`edgedb instance logs -I movies`
starting PostgreSQL 14.4 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 10.2.1 20210130 (Red Hat 10.2.1-11), 64-bit
listening on Unix socket "/run/user/1000/edgedb-movies/.s.PGSQL.5432"
Ah.... I got it from the process info, your bundled Postgres is here
`/home/athan/.local/share/edgedb/portable/2.1/bin/`
ii. Is it possible to use my local Postgres server installation with EdgeDB and view tables there ?
So I guess the quick answer here is no :-)
iii. Can you explain what is the purpose of having multiple databases in the same instance instead of creating separate instance-database pairs ? Is it for exchanging, sharing data more easily between two different databases ?
iv. Is it possible to access your bundled Postgres server and read data (`/home/athan/.local/share/edgedb/data/`) stored in tables from some other client, database administration tool e.g. DBeaver ?

5. memory - buffers - cache
How can I read the default values for the following configuration settings ?
i. `effective_cache_size -> cfg::memory`
An estimate of the effective size of the disk cache available to a single query.
ii. `shared_buffers -> cfg::memory`
The amount of memory used for shared memory buffers.
iii. `query_work_mem -> cfg::memory`
The amount of memory used by internal query operations such as sorting.
iv. What is this Cache  located in ` /home/athan/.cache/edgedb/ ` ?
Can you elaborate more on their usage/differences and point to some examples ?

PS: Kudos to EdgeDB team, the query language and the schema definition language is perhaps the best of the best I have seen so far in more than 10 years of continuous exploration of every DBMS and data engine in IT industry.

**Comment by tailhook** [comment, marked as answer]
> what is the default password ?

There is no default password. Unique password is generated for each instance, and it is stored in `~/.config/edgedb/credentials/<instance-name>.json`. But it's not expected to look into that file manually. `edgedb -I <instance-name>` or your app running from the project dir is expected to find those credentials automatically.

> How can I change the path of the data-dir for the instance ?

We don't have this kind of configuration in the CLI. When running server manually use `--data-dir` command-line option.

> How can I change the auto-shutdown-after ?

Same thing here for now.

We can add some option, but we are not sure if we want to persist them and how. Or just use as one-time thing mostly in combination with `--foreground`.

> iii. What is the analogy of instances with Postgres here ?

Entirely separate database server running from a separate data directory.

> Is this instance a Postgres process that is forked from the server ?

No. Instances are independent of each other. Each instance has its own data dir, postgres server, and other auxilliary processes if needed.

> Is it possible to relocate data to a different storage device and/or specify different data folder for the server ?

We don't have a specific option for that. But you can use symlinks or mountpoints to refer to a different storage device.

> Is it possible to use my local Postgres server installation with EdgeDB and view tables there ?

You can use the server. But we have our own way of handling tables (in short: tables and columns are denoted as uuids instead of names internally). So you can't access data created by the different means using EdgeDB.

> Can you explain what is the purpose of having multiple databases in the same instance instead of creating separate instance-database pairs ? Is it for exchanging, sharing data more easily between two different databases ?

Not too much actually, but here are some ideas:
1. To put some quick experiment in the REPL
2. To have test database in the same instance, or even multiple test databases
Instances are more memory-intensive so using many of them can be costly.

**Edit:** I don't think we have any way of "sharing data more easily between two different databases". The only thing that is shared is credentials. But even then you need to create a new connection to access a different database, although you can use `\c` shortcut within the REPL (So don't have to restart the REPL).

Generally you could even put different application into different databases of the single instance. But this is not convenient for the development because different applications usually have different upgrade cycle, so eventually you need a separate version of EdgeDB for each app (even if it's just for a small time period).

> Is it possible to access your bundled Postgres server and read data (/home/athan/.local/share/edgedb/data/) stored in tables from some other client, database administration tool e.g. DBeaver ?

Those tables have cryptic names. We have a plan to make a proxy that would be able to parse SQL and translate human-friendly names EdgeDB exposes to the users into internal names, so you could use SQL and your favourite tool on EdgeDB data. But I don't think we have a deadline for that feature.

> How can I read the default values for the following configuration settings ?

```edgedb> select cfg::Config { effective_cache_size, shared_buffers, query_work_mem };
{cfg::Config {effective_cache_size: <cfg::memory>'4GiB', shared_buffers: <cfg::memory>'128MiB', query_work_mem: <cfg::memory>'4MiB'}}
```

> What is this Cache located in /home/athan/.cache/edgedb/ ?

Downloads of the EdgeDB packages and version checks currently. We may add more things there, but nothing related to the application's data will be put there in the foreseeable future. It's just a stash dir for the CLI.
